Added transcendental checks, rewrite exp for power

sidhantnagpal opened the pull request:
#### References to other Issues or PRs
Fixes #14262 
Fixes #12744

#### Brief description of what is fixed or changed
`_eval_rewrite_as_exp` is added for Pow


#### Other Comments
Fixed a couple of bugs in `solvers` with help from @jksuom @smichr 
Fixes #14791

Commit:
added transcendental, rewrite exp

sidhantnagpal commented:
Looks like an issue with transcendental (travis). Will look into it soon.

asmeurer commented on sympy/core/power.py:
```diff
@@ -1212,6 +1212,22 @@ def _is_one(expr):
+
+        if self.base.is_zero or _is_one(self.base):
+            return False
+        elif self.base.is_real and self.exp.is_irrational:
```
According to the theorem, this should be `is_algebraic`, not `is_real`.

sidhantnagpal replied:
Yes, the issue is the theorem happens to use a different meaning for algebraic (including complex values). [Here](https://en.wikipedia.org/wiki/Gelfond%E2%80%93Schneider_theorem).

asmeurer commented on sympy/core/power.py:
```diff
@@ -1212,6 +1212,22 @@ def _is_one(expr):
+            return False
+        elif self.base.is_real and self.exp.is_irrational:
+            return True
+        elif self.base.is_transcendental and self.exp.is_zero == False:
```
This is also wrong. The theorem only applies when the base is algebraic (see https://en.wikipedia.org/wiki/Gelfond%E2%80%93Schneider_theorem, the counterexample `(sqrt(2)**sqrt(2))**sqrt(2)`. This would be true if you added `self.exp.is_rational` as `t**r` is transcendental for transcendental `t` and nonzero rational `r`.

sidhantnagpal replied:
Yes, I also figured this, will make the required changes in next round. Thanks,

asmeurer commented on sympy/core/power.py:
```diff
@@ -1212,6 +1212,22 @@ def _is_one(expr):
                 return self.exp.is_rational
 
+    def _eval_is_transcendental(self):
+        # Gelfond-Schneider theorem
```
Isn't this already implemented in `_eval_is_algebraic`? The assumptions system already knows that `transcendental == complex & !algebraic`.

sidhantnagpal replied:
Where is this exactly?
```
>>> (4 + 5*I).is_transcendental
False
```
Complex values with rational real/imag give False in current master.

sidhantnagpal replied:
Yes, looks like the remaining values are correct in master.
```
>> ((S(2)/3 + I*S(7)/8)**sqrt(2)).is_transcendental
True
>> ((1 + sqrt(3))**sqrt(2)).is_transcendental
True
```
So, retaining transcendental for exp for the time being.

Commit:
remove transcendental

sidhantnagpal commented:
Looks like `_tsolve` and `_solve` are leading to recursion depth exceeded.
What could be the reason for it?

sidhantnagpal reviewed (commented):
Although a minor issue, but is there a reason the test file was named `test_eval_power.py` rather than `test_power.py`.

asmeurer commented:
I don't know. You might find the answer if you look through the git history.

jksuom commented:
The file was created by commit dfc19f5189 whose message says: Fixed bug in Rational._eval_power and added preliminary tests for this type. So it seems that initially only _eval_power was tested.

sidhantnagpal commented:
Okay, if renaming is to be done, I can include it in this PR itself.

sidhantnagpal commented:
Okay, so the recursion error is due to the [line](https://github.com/sympy/sympy/blob/master/sympy/solvers/solvers.py#L2678) `rewrite = lhs.rewrite(exp)` in `_tsolve` (transcendental solve). An alternate might be required for the intended use.

sidhantnagpal commented:
Yes. There seems to be following issues.
For `rewrite`:
`expand_complex(x)` will cause variable expansion to `re(x) + I*im(x)` (under default assumptions, which can cause problems in `solvers` where variables should remain atomic).

For `exp`:
Should `exp(x*log(32))` automatically reduce to `32*exp(x)`?
(In general `exp(expr1 * log(expr2) * log(exp3) * expr4)` to `expr2 * expr3 * exp(expr1 * expr4)`)

For `_tsolve`:
The statement `lhs.rewrite(exp)` should be rethought on possibly to handle rewrites which change the expression format (the initial intention for writing it may no longer hold).

asmeurer commented:
Why is expand_complex used at all? Should it rather be expand_complex(deep=False)?

sidhantnagpal commented:
I thought about using deep=False, but unfortunately it still expands variables
```
>>> x = Symbol('x')
>>> expand_complex(x, deep=False)
re(x) + I*im(x)
```

sidhantnagpal commented:
> In fact I would hard-code it

Yes, I have hard coded it as used in the snippet [here](https://github.com/sympy/sympy/pull/14712#issuecomment-389757705).

sidhantnagpal commented:
Regarding the solvers issue:
For the test `solve(3*x + 5 + 2**(-5*x + 3), x)`:
(adding `print(lhs, rhs, rewrite, sep=' | ')` after [`rewrite = lhs.rewrite(exp)`](https://github.com/sympy/sympy/blob/master/sympy/solvers/solvers.py#L2678))

The previous results are:
```
2**(-5*x + 3) + 3*x | -5 | 2**(-5*x + 3) + 3*x
32**_x*(3*_x + 5) | -8 | 32**_x*(3*_x + 5)
```

The current results are:
```
2**(-5*x + 3) + 3*x | -5 | 3*x + exp((-5*x + 3)*log(2))
2**(-5*x + 3) + 3*x | -5 | 3*x + exp((-5*x + 3)*log(2))
2**(-5*x + 3) + 3*x | -5 | 3*x + exp((-5*x + 3)*log(2))
... (infinite times)
```

What exactly does the rewrite line do?

sidhantnagpal commented:
The tests affected by `lhs.rewrite(exp)` are:
`solve(sin(x) + tan(x))` and `solve(sin(3*x) + sin(6*x))`.

jksuom commented:
It seems that test_issue_14640 repeatedly tries to solve (or `_tsolve`) the same equation. There is a flag, `flags['tsolve_saw']` that collects those equations that have been examined, but that is not passed on [here](https://github.com/sympy/sympy/blob/master/sympy/solvers/solvers.py#L2640),

sidhantnagpal commented:
Yes, this fixes the infinite recursion.
Although, there are still failing tests due to the rewrite (as more equations are now in the scope of `lhs!=rewrite`). I will push the changes for further review.

Commit:
improve rewrite exp

Commit:
add test_power file

asmeurer commented on sympy/core/power.py:
```diff
@@ -1232,6 +1233,17 @@ def _eval_is_algebraic_expr(self, syms):
+                # (as exp(x*log(5)) automatically reduces to x**5)
+                return exp(log(base)*expo, evaluate=expo.has(Symbol))
+            else:
+                u, v = log(base).as_real_imag()
```
I would still do deep=False here, or just use `log(abs(base)) + arg(base)*I`. rewrite shouldn't do deep changes to the expression that are unrelated to the function it is rewriting to.

sidhantnagpal replied:
Okay, will make the changes along with the remaining tests.

sidhantnagpal commented:
I tried to trace the execution for the failing tests. Turns out all of them are related to solvers.
The failing tests are:
1. test_util.py: 
`function_range(tan(x), x, Interval(0, pi))`
`continuous_domain(tan(x), x, Interval(0, 2*pi))`
2. test_inequalities.py:
`solve_univariate_inequality(tan(x) < S.One, x, relational=False)`
3. test_solvers.py:
`solve(3*log(a**(3*x + 5)) + a**(3*x + 5), x)`
`solve(sin(x) + tan(x))`

I am not that versed in solvers. It would be better if someone could help with these tests. 
Meanwhile, the changes to power.py are almost complete.

Commit:
use arg, abs in rewrite exp

jksuom commented:
`solveset` will use `rewrite(exp)` for trigonometric functions. When trying to find the zeros of `1/tan(x)` it will get
```
>>> (1/tan(x)).rewrite(exp)
exp(-log(I*(-exp(I*x) + exp(-I*x))/(exp(I*x) + exp(-I*x))))
```
instead of
```
>>> (1/tan(x)).rewrite(exp)
-I*(exp(I*x) + exp(-I*x))/(-exp(I*x) + exp(-I*x))
```
Apparently it cannot solve the modified form and returns:
```
>>> solveset(1/tan(x), x)
EmptySet()
```
instead of
```
>>> solveset(1/tan(x), x)
Union(ImageSet(Lambda(_n, 2*_n*pi + pi/2), S.Integers), ImageSet(Lambda(_n, 2*_n*pi + 3*pi/2), S.Integers))
```

sidhantnagpal commented:
Also, in current master:
`solve(exp(log(5)*x) - 2**x, x)` gives `[0]`
but
`solve(exp(log(5)*x) - exp(log(2)*x), x)` gives `[]`.

Commit:
detect exp in rewrite

sidhantnagpal commented:
`test_util` and `test_inequalities` are now fixed.

`test_solvers` still has the following failing tests:
`solve(exp(log(5)*x) - 2**x, x)` and `solve(3*log(a**(3*x + 5)) + a**(3*x + 5), x)`.

Commit:
remove newline

smichr commented:
> solve(exp(log(5)*x) - exp(log(2)*x), x) gives []

This gives [0] if x is declared as real.

sidhantnagpal commented:
Yes, that helps, but it would be good to have so under default assumptions (without knowing the nature of solution).

smichr commented:
There is a genuine problem in the code. Let me see if I can do a pr against this branch.

smichr committed:
fix _invert and modify tests

sidhantnagpal commented:
Sure, that'd be great.

smichr commented:
OK, I sent the PR. A good side effect is that all 3 solutions are now given for a generic variable (you will see the modified test in the test file).

smichr committed:
handle f(x) = f(1) sort of cases

The upper code handles cases where the two functions
share a symbol of interest; the lower handles the
case where they don't.

smichr commented:
Until you press the Commit button on the PR that I sent, the changes won't appear in *this* branch.

sidhantnagpal commented:
Yes, everything is working fine. Merging it. Thanks.

Commit:
Merge pull request #1 from smichr/power-exp

fix _invert and modify tests

smichr commented:
it's in

sidhantnagpal commented:
Great, this also closes #14791
